From: Didier Raboud <odyx@debian.org>
Date: Tue, 25 Aug 2020 22:12:28 +0200
Subject: Fix att{t,}ributes typo

---
 cupsfilters/ipp.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/cupsfilters/ipp.c b/cupsfilters/ipp.c
index 53bb531..ed2745e 100644
--- a/cupsfilters/ipp.c
+++ b/cupsfilters/ipp.c
@@ -349,7 +349,7 @@ get_printer_attributes5(http_t *http_printer,
       /* Log all printer attributes for debugging and count them */
       if (debug)
 	log_printf(get_printer_attributes_log,
-		   "Full list of all IPP atttributes:\n");
+		   "Full list of all IPP attributes:\n");
       attr = ippFirstAttribute(response);
       while (attr) {
 	total_attrs ++;
