From c7a14fe53efd1bfdc682556bca5a4a3931372d84 Mon Sep 17 00:00:00 2001
From: Didier Raboud <odyx@debian.org>
Date: Mon, 5 Feb 2018 08:37:45 +0100
Subject: Fix Ou{,t}put typo in foomatic-rip

---
 filter/foomatic-rip/foomaticrip.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/filter/foomatic-rip/foomaticrip.c b/filter/foomatic-rip/foomaticrip.c
index 2a642edaf..22e6dff15 100644
--- a/filter/foomatic-rip/foomaticrip.c
+++ b/filter/foomatic-rip/foomaticrip.c
@@ -1075,7 +1075,7 @@ int main(int argc, char** argv)
         dstrclear(postpipe);
 
     if (postpipe->len)
-        _log("Ouput will be redirected to:\n%s\n", postpipe);
+        _log("Output will be redirected to:\n%s\n", postpipe);
 
 
     filename = strtok_r(filelist->data, " ", &p);
