Description: Fix PDF landscape printing
Bug-Redhat: https://bugzilla.redhat.com/show_bug.cgi?id=768811
Author: Tim Waugh <twaugh@redhat.com>
Last-Update: 2013-11-28
diff -up cups-filters-1.0.40/filter/pdftops.c.pdf-landscape cups-filters-1.0.40/filter/pdftops.c
--- cups-filters-1.0.40/filter/pdftops.c.pdf-landscape	2013-10-11 17:15:45.639623613 +0100
+++ cups-filters-1.0.40/filter/pdftops.c	2013-10-11 17:22:41.517421905 +0100
@@ -273,7 +273,6 @@ main(int  argc,				/* I - Number of comm
 		*pstops_options,	/* Options for pstops filter */
 		*pstops_end;		/* End of pstops filter option */
   const char	*cups_serverbin;	/* CUPS_SERVERBIN environment variable */
-  const char	*content_type;		/* CONTENT_TYPE environment variable */
 #if defined(HAVE_SIGACTION) && !defined(HAVE_SIGSET)
   struct sigaction action;		/* Actions for POSIX signals */
 #endif /* HAVE_SIGACTION && !HAVE_SIGSET */
@@ -457,7 +456,6 @@ main(int  argc,				/* I - Number of comm
   * Build the command-line for the pdftops or gs filter...
   */
 
-  content_type = getenv("CONTENT_TYPE");
   if (renderer == PDFTOPS)
   {
     pdf_argv[0] = (char *)"pdftops";
@@ -646,25 +644,17 @@ main(int  argc,				/* I - Number of comm
     else if ((renderer == PDFTOPS) || (renderer == PDFTOCAIRO))
     {
      /*
-      *  Use the page sizes of the original PDF document, this way documents
-      *  which contain pages of different sizes can be printed correctly
+      *  No options to add. Note: pdftopdf has already rotated pages
+      *  for us.
       */
-
-      /* Only do this for unprocessed PDF files */
-      if (content_type && !strstr (content_type, "/vnd.cups-"))
-        pdf_argv[pdf_argc++] = (char *)"-origpagesizes";
     }
 #endif /* HAVE_POPPLER_PDFTOPS_WITH_ORIGPAGESIZES */
     else if (renderer == ACROREAD)
     {
      /*
-      * Use the page sizes of the original PDF document, this way documents
-      * which contain pages of different sizes can be printed correctly
+      *  No options to add. Note: pdftopdf has already rotated pages
+      *  for us.
       */
-     
-      /* Only do this for unprocessed PDF files */
-      if (content_type && !strstr (content_type, "/vnd.cups-"))
-        pdf_argv[pdf_argc++] = (char *)"-choosePaperByPDFPageSize";
     }
 
    /*
