Description: Include dbus so that colord support works
Bug-Redhat: https://bugzilla.redhat.com/show_bug.cgi?id=1026928
Author: Tim Waugh <twaugh@redhat.com>
Last-Update: 2013-11-28
diff -up cups-filters-1.0.41/Makefile.am.dbus cups-filters-1.0.41/Makefile.am
--- cups-filters-1.0.41/Makefile.am.dbus	2013-10-25 14:12:56.000000000 +0100
+++ cups-filters-1.0.41/Makefile.am	2013-11-06 17:16:58.504234619 +0000
@@ -474,9 +474,11 @@ gstoraster_SOURCES = \
 	cupsfilters/raster.h
 gstoraster_CFLAGS = \
 	$(CUPS_CFLAGS) \
+	$(DBUS_CFLAGS) \
 	-I$(srcdir)/cupsfilters/
 gstoraster_LDADD = \
 	$(CUPS_LIBS) \
+	$(DBUS_LIBS) \
 	libcupsfilters.la
 
 imagetopdf_SOURCES = \
diff -up cups-filters-1.0.41/config.h.in.dbus cups-filters-1.0.41/config.h.in
diff -up cups-filters-1.0.41/configure.ac.dbus cups-filters-1.0.41/configure.ac
--- cups-filters-1.0.41/configure.ac.dbus	2013-10-30 07:21:44.000000000 +0000
+++ cups-filters-1.0.41/configure.ac	2013-11-06 17:23:07.619921079 +0000
@@ -114,6 +114,18 @@ AX_COMPARE_VERSION([$CUPS_VERSION],[gt],
 AC_DEFINE(PDFTOPDF, [], [Needed for pdftopdf filter compilation])
 AC_DEFINE_DIR(BANNERTOPDF_DATADIR, "{CUPS_DATADIR}/data", [Directory where bannertopdf finds its data files (PDF templates)])
 
+dnl DBus support
+DBUS_CFLAGS=""
+DBUS_LIBS=""
+AC_ARG_ENABLE([dbus], AC_HELP_STRING([--disable-dbus],
+    [Do not use dbus to communicate with external services]))
+if test "$enable_dbus" != "no"; then
+        PKG_CHECK_MODULES(DBUS, [dbus-1], [HAVE_DBUS=-DHAVE_DBUS])
+        AC_DEFINE([HAVE_DBUS], [], [Defines if dbus is available.])
+fi
+AC_SUBST(DBUS_CFLAGS)
+AC_SUBST(DBUS_LIBS)
+
 AC_SEARCH_LIBS([dlopen],
 	[dl],
 	[AS_IF([test "$ac_cv_search_dlopen" != "none required"], [
@@ -640,6 +652,7 @@ Build configuration:
 	php-config:      ${with_php_config}
 	test-font:       ${with_test_font_path}
 	tiff:            ${with_tiff}
+	dbus:            ${HAVE_DBUS}
 	avahi:           ${enable_avahi}
 	browsing:        ${with_browseremoteprotocols}
 	werror:          ${enable_werror}
diff -up cups-filters-1.0.41/filter/colord.c.dbus cups-filters-1.0.41/filter/colord.c
--- cups-filters-1.0.41/filter/colord.c.dbus	2013-07-29 20:12:42.000000000 +0100
+++ cups-filters-1.0.41/filter/colord.c	2013-11-06 17:16:58.505234623 +0000
@@ -32,6 +32,8 @@ MIT Open Source License  -  http://www.o
 #include <stdio.h>
 #include <sys/types.h>
 
+#include "config.h"
+
 #ifdef HAVE_DBUS
   #include <dbus/dbus.h>
 #endif
